srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,957 @@
+			ensure_root(origin)?;
+			<TermDuration<T>>::put(count);
+		}
+
+		fn on_initialize(n: T::BlockNumber) {
+			if let Err(e) = Self::end_block(n) {
+				runtime_io::print("Guru meditation");
+				runtime_io::print(e);
+			}
+		}
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where <T as system::Trait>::AccountId {
+		/// A new term with new members.
+		NewTerm(Vec<AccountId>),
+		/// No candidates were elected for this round.
+		EmptyCouncil(),
+	}
+);
+
+impl<T: Trait> Module<T> {
+	/// Check if `who` is a candidate. It returns the insert index if the element does not exists as
+	/// an error.
+	fn is_candidate(who: &T::AccountId) -> Result<(), usize> {
+		Self::candidates().binary_search(who).map(|_| ())
+	}
+
+	/// Check if `who` is a voter in the ongoing round.
+	fn is_current_voter(who: &T::AccountId) -> bool {
+		let now = Self::election_rounds();
+		Self::votes_of(who).1 == now && <StakeOf<T>>::exists(who)
+	}
+
+	/// Check if `who` was a voter who has not yet called `remove_voter`.
+	fn is_old_voter(who: &T::AccountId) -> bool {
+		let now = Self::election_rounds();
+		Self::votes_of(who).1 < now && <StakeOf<T>>::exists(who)
+	}
+
+	/// The locked stake of a voter.
+	fn locked_stake_of(who: &T::AccountId) -> BalanceOf<T> {
+		if Self::is_current_voter(who) {
+			Self::stake_of(who)
+		} else {
+			Zero::zero()
+		}
+	}
+
+	/// Check there's nothing to do this block.
+	///
+	/// Runs phragmen election and cleans all the previous candidate state. The voter state is NOT
+	/// cleaned and voters must themselves submit a transaction to clean it.
+	fn end_block(block_number: T::BlockNumber) -> dispatch::Result {
+		if (block_number % Self::term_duration()).is_zero() {
+			use sr_primitives::phragmen;
+			let seats = Self::desired_seats() as usize;
+			let candidates = Self::candidates();
+			let current_round = Self::election_rounds();
+			let voters_and_votes = <VotesOf<T>>::enumerate()
+				.filter_map(|(v, i)| if i.1 == current_round { Some((v, i.0)) } else { None } )
+				.collect::<Vec<(T::AccountId, Vec<T::AccountId>)>>();
+
+			let maybe_new_members = phragmen::elect::<_, _, _, T::CurrencyToVote>(
+				seats,
+				false,
+				0,
+				candidates.clone(),
+				voters_and_votes,
+				Self::locked_stake_of
+			);
+
+			let old_members = <Members<T>>::take();
+			if let Some((mut new_members_with_approval, _)) = maybe_new_members {
+				new_members_with_approval.sort_by_key(|x| x.0.clone());
+
+				let new_members = new_members_with_approval
+					.into_iter()
+					.filter_map(|(m, a)| if !a.is_zero() { Some(m) } else { None } )
+					.collect::<Vec<T::AccountId>>();
+				<Members<T>>::put(new_members.clone());
+
+				// return bond to losers.
+				candidates.into_iter().for_each(|c| {
+					// if this candidate is a loser return the bond
+					if new_members.binary_search(&c).is_err() {
+						T::Currency::unreserve(&c, T::CandidacyBond::get());
+					}
+				});
+
+				T::ChangeMembers::change_members(
```

Comment:
Incorrect. Use `T::ChangeMembers::set_members_sorted` if you don't know who the actual incoming and outgoing members are.

Reply:
But I don't understand from reading the docs of `trait ChangeMemeber` why in if a single one is kicked out `T::ChangeMembers::change_members(&[], &[who], members);` is fine? Based on your comment, shouldn't all of them be `set_members_sorted`?

Reply:
they're just meant for different situations. if you naturally have the diff ahead of time (like elections module does), then you can avoid some computation and provide it directly. otherwise, you pass the pre-sorted set and have it figure out the diff for you.